crates/solidity/inputs/schema/grammar/05-expressions/01-base-expressions/productions.yml

```
@@ -138,8 +138,17 @@
             close:
               reference: "CloseBracket"
 
-      primaryExpressions:
-        - reference: "PrimaryExpression"
+      primaryExpression:
+        choice:
+          - reference: "Identifier"
```

Not sure I understand why this is removed and duplicated in the parent base expression? I was hoping to refactor the other expressions in the future to define them in a single place, so that:

1. The definition is not duplicated, and harder to diverge/introduce bugs in the future.
2. It is easier to figure out what changed between versions, if the definition is compact.

```suggestion
      primaryExpression:
        reference: "PrimaryExpression"
```

---

Because IIRC you wanted this as a single ParserRef for precisely this reason i.e. to not have a PrimaryExpression wrapper.

IMO the tree is better without PrimaryExpression, because it is an artifact of the grammar that serves no useful purpose. Better to find a different way to deal with those duplication concerns that doesn't affect the tree we produce. Maybe once we have `inline`.

---

I wonder if you mean [this comment](https://github.com/NomicFoundation/slang/pull/332#discussion_r1119430925)?

> What is the use case for the array of references vs a single reference? Is `primary_expressions` array intended to be alternatives?
> 
> I see the grammar still defines `PrimaryExpression` as a single `reference:` to a `choice:`, instead of inlining them in the parent operator definition, which is great, as it doesn't duplicate its contents between multiple versions, and doesn't introduce new types/structures to the schema.
> 
> Should we change it to a single primary expression in this case?

I agree we shouldn't have `PrimaryExpression` in the tree, and this is one of many cases `inline` would solve. But if you agree, I suggest leaving this to beta/later, as the deduplicating makes it easier for refactoring/diffing versions. I imagine a lot of them will be introduced by #340

---

ok